game/system/newpad.cpp

```diff
@@ -113,66 +149,114 @@ int IsPressed(MappingInfo& mapping, Button button, int pad = 0) {
+        } else {
+          input =
+              input -
+              g_buffered_key_status[g_mapping.keyboard_analog_mapping[pad][analog].negative_key];
+        }
+      }
+      g_key_analogs[pad][analog] = input;
+    }
+  }
+}
+
 // returns the value of the analog axis (in the future, likely pressure sensitive if we support it?)
 // if invalid or otherwise -- returns 127 (analog stick neutral position)
-int AnalogValue(MappingInfo& /*mapping*/, Analog analog, int pad = 0) {
+int GetAnalogValue(MappingInfo& /*mapping*/, Analog analog, int pad = 0) {
+  float input = 0.0f;
   if (CheckPadIdx(pad) == -1) {
     // Pad out of range, return a stable value
     return 127;
   }
 
-  float input = 0.0f;
-  if (pad == 0) {
-    // Movement controls mapped to WASD keys
-    if (g_buffered_key_status[GLFW_KEY_W] && analog == Analog::Left_Y)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_S] && analog == Analog::Left_Y)
-      input += 1.0f;
-    if (g_buffered_key_status[GLFW_KEY_A] && analog == Analog::Left_X)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_D] && analog == Analog::Left_X)
-      input += 1.0f;
-
-    // Camera controls mapped to IJKL keys
-    if (g_buffered_key_status[GLFW_KEY_I] && analog == Analog::Right_Y)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_K] && analog == Analog::Right_Y)
-      input += 1.0f;
-    if (g_buffered_key_status[GLFW_KEY_J] && analog == Analog::Right_X)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_L] && analog == Analog::Right_X)
-      input += 1.0f;
-  } else if (pad == 1) {
-    // these bindings are not sane
-    if (g_buffered_key_status[GLFW_KEY_KP_5] && analog == Analog::Left_Y)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_KP_2] && analog == Analog::Left_Y)
-      input += 1.0f;
-    if (g_buffered_key_status[GLFW_KEY_KP_1] && analog == Analog::Left_X)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_KP_3] && analog == Analog::Left_X)
-      input += 1.0f;
-
-    // these bindings are not sane
-    if (g_buffered_key_status[GLFW_KEY_KP_DIVIDE] && analog == Analog::Right_Y)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_KP_8] && analog == Analog::Right_Y)
-      input += 1.0f;
-    if (g_buffered_key_status[GLFW_KEY_KP_7] && analog == Analog::Right_X)
-      input += -1.0f;
-    if (g_buffered_key_status[GLFW_KEY_KP_9] && analog == Analog::Right_X)
-      input += 1.0f;
-  }
-
-  if (input == 0) {
+  if (g_gamepads.gamepad_idx[pad] == -1) {
+    input = g_key_analogs[pad][(int)analog];
+    // Hack. Clearing the buffer immediately can lead to inconsistencies on analog input.
+    // If a mouse is disconnected or can't calculate a new delta it would stay stuck at 1.
+    // Decreasing the values gradually seems like a good comprise.
+    g_key_analogs[pad][(int)analog] *= 0.95;
+  } else {  // Gamepad present
```

Comment:
Just for clarification. Are you expecting something like this in the code.

```cpp
  float input = 0.0f;
  if (CheckPadIdx(pad) == -1) {
    // Pad out of range, return a stable value
    return 127;
  }

  if (g_gamepads.gamepad_idx[pad] > -1) {
    input += g_gamepad_analogs[pad][(int)analog];
  }
  input += g_key_analogs[pad][(int)analog];
  // Hack. Clearing the buffer immediately can lead to inconsistencies on analog input.
  // If a mouse is disconnected or can't calculate a new delta it would stay stuck at 1.
  // Decreasing the values gradually seems like a good comprise.
  g_key_analogs[pad][(int)analog] *= 0.95;
```

Reply:
Sort of. You can track key and controller analogs separately (and thus decay keyboard just fine), but when asked for analog value you pick whichever one is "most active."

Reply:
So something like this?

```cpp
  float input = 0.0f;
  if (CheckPadIdx(pad) == -1) {
    // Pad out of range, return a stable value
    return 127;
  }

 float controller_input, keyboard_input;
controller_input = keyboard_input = 0.0f;
  if (g_gamepads.gamepad_idx[pad] > -1) {
    controller_input = g_gamepad_analogs[pad][(int)analog];
  }
  keyboard_input = g_key_analogs[pad][(int)analog];
  g_key_analogs[pad][(int)analog] *= 0.95;

if(fabs(controller_input) > fabs(keyboard_input)){
   input = controller_input;
} else {
   input = keyboard_input;
}
 glfw_to_goal_value(input); //pseudo code

```

Reply:
Yeah, that's sort of what I was thinking as well.